File: src/main/java/dev/kensa/render/diagram/directive/UmlBox.java
package dev.kensa.render.diagram.directive;

import java.util.List;

import static dev.kensa.util.WrappingCollector.wrapping;
import static java.lang.String.format;
import static java.util.Arrays.asList;

public class UmlBox implements UmlDirective {
    public static UmlBox surroundingBox(UmlParticipant... participants) {
        return new UmlBox(participants);
    }

    public static UmlBox surroundingBox(String title, UmlParticipant... participants) {
        return new UmlBox(participants).withTitle(title);
    }

    private final List<UmlParticipant> participants;
    private final String title;
    private final String colour;

    private UmlBox(UmlParticipant... participants) {
        this(asList(participants));
    }

    private UmlBox(List<UmlParticipant> participants) {
        this(participants, "", "");
    }

    private UmlBox(List<UmlParticipant> participants, String title, String colour) {
        this.participants = participants;
        this.title = title;
        this.colour = colour;
    }

    public UmlBox withTitle(String title) {
        return new UmlBox(participants, title, this.colour);
    }

    public UmlBox withColour(String colour) {
        return new UmlBox(participants, this.title, colour);
    }

    @Override
    public List<String> asUml() {
        return participants.stream().flatMap(p ->p.asUml().stream()).collect(wrapping(buildTitle(), "end box"));
    }

    private String buildTitle() {
        return format("box %s %s", title.isEmpty() ? "" : "\"" + title + "\"", colour).replaceAll("\\s+", " ").trim();
    }

}

File: core/src/main/kotlin/dev/kensa/render/diagram/directive/UmlBox.kt
package dev.kensa.render.diagram.directive

class UmlBox private constructor(private val participants: List<UmlParticipant>, private val title: String = "", private val colour: String = "") : UmlDirective {
    private constructor(vararg participants: UmlParticipant) : this(listOf(*participants))

    fun withTitle(title: String): UmlBox = UmlBox(participants, title, colour)

    fun withColour(colour: String): UmlBox = UmlBox(participants, title, colour)

    override fun asUml(): List<String> = listOf(buildTitle()) + participants.flatMap { it.asUml() } + "end box"

    private fun buildTitle(): String = "box ${title.takeIf { it.isNotEmpty() }?.let { "\"$it\"" }.orEmpty()} $colour".replace("\\s+".toRegex(), " ").trim()

    companion object {
        @JvmStatic
        fun surroundingBox(vararg participants: UmlParticipant): UmlBox = UmlBox(*participants)

        @JvmStatic
        fun surroundingBox(title: String, vararg participants: UmlParticipant): UmlBox = UmlBox(*participants).withTitle(title)
    }
}
